src/transformStateWithClones.js

```diff
@@ -7,7 +7,36 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const resultArray = [];
+
+  for (const action of actions) {
+    const { type } = action;
+    const currentState = { ...resultArray[resultArray.length - 1] || state };
+
+    if (type === 'addProperties') {
+      const { extraData } = action;
+
+      Object.assign(currentState, extraData);
+    } else if (type === 'removeProperties') {
+      const { keysToRemove } = action;
+
+      for (const key of keysToRemove) {
+        if (currentState.hasOwnProperty(key)) {
+          delete currentState[key];
+        }
+      }
+    } else if (type === 'clear') {
+      for (const key in currentState) {
```

Comment:
There's no need to check if a property belongs to an object before deleting it. The delete operator is safe to use regardless, because it doesn't throw an error if the property doesn't exist.